### ISSUE TITLE
authorize OSM access is not working after change to external browser

### ISSUE THREAD

**Issue by matkoniecz** [westnordost/StreetComplete#386]
on uninstall + install of 75e86f1:

settings -> authorize OSM access -> select firefox (54.0.1) -> authorize on osm page (I was already logged in in firefox)

expected: error with explanation in app or authorization success

what happens: I am back in settings menu with "Not yet authorized" in "authorize OSM access" in Android Monitor once following exception appeared (in other cases no log that seemed appeared):

```
07-07 12:55:37.020 5634-5634/de.westnordost.streetcomplete E/OsmOAuthDialogFragment: Error during authorization
                                                                                     oauth.signpost.exception.OAuthNotAuthorizedException: Authorization failed (server replied with a 401). This can happen if the consumer key was not correct or the signatures did not match.
                                                                                         at oauth.signpost.AbstractOAuthProvider.handleUnexpectedResponse(AbstractOAuthProvider.java:243)
                                                                                         at oauth.signpost.AbstractOAuthProvider.retrieveToken(AbstractOAuthProvider.java:193)
                                                                                         at oauth.signpost.AbstractOAuthProvider.retrieveAccessToken(AbstractOAuthProvider.java:108)
                                                                                         at de.westnordost.streetcomplete.oauth.OsmOAuthDialogFragment$RetrieveAccessTokenTask.doInBackground(OsmOAuthDialogFragment.java:235)
                                                                                         at de.westnordost.streetcomplete.oauth.OsmOAuthDialogFragment$RetrieveAccessTokenTask.doInBackground(OsmOAuthDialogFragment.java:231)
                                                                                         at de.westnordost.streetcomplete.util.InlineAsyncTask.doInBackground(InlineAsyncTask.java:24)
                                                                                         at de.westnordost.streetcomplete.util.InlineAsyncTask.doInBackground(InlineAsyncTask.java:11)
                                                                                         at android.os.AsyncTask$2.call(AsyncTask.java:295)
                                                                                         at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                                         at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
                                                                                         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                                         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                                         at java.lang.Thread.run(Thread.java:818)
```

**Comment by westnordost** [comment]
In the case of error during authorization, a Toast should pop up, saying that authorization failed or something. Did it not pop up?

With uninstall + install you mean "clear data"? Can you reproduce it somehow?

**Comment by matkoniecz** [comment]
Toast failed to appear. My phone claim that uninstalling clears all data associated with application (and from what I noticed it seems to be true).

**Comment by matkoniecz** [comment]
Toast appeared after minimizing SC and switching back to it. I will make recording showing the problem.

**Comment by matkoniecz** [comment]
now it is behaving differently - it is going back to SC without any action (probably FF cached something - I will post this variant of bug now and try to reproduce original behavior where I interacted with OSM website).

![remove](https://user-images.githubusercontent.com/899988/27983063-387eccd0-63b4-11e7-94a7-ed2d90eff590.gif)
![bug](https://user-images.githubusercontent.com/899988/27983101-0b1ff8b2-63b5-11e7-887b-8d42a77ae67e.gif)

**Comment by krzyk** [comment]
I also can't authorize using Firefox. I thought that it's a problem with tabs queue (Firefox allows the user to open the links after some time) - it made SC right away show unauthorized.
But when I disabled that feature and tried Firefox once more, I authorized in the browser and when I returned to SC it still showed that I'm not authorized.

When I changed to Chrome it worked.

**Comment by matkoniecz** [comment]
![ezgif-4-db4707f476](https://user-images.githubusercontent.com/899988/27983262-989384e0-63b8-11e7-9764-2bc30ebd1102.gif)

**Comment by matkoniecz** [comment]
For me Chrome workaround also works.

**Comment by matkoniecz** [comment]
Yandex Browser also failed.

**Comment by matkoniecz** [comment]
Reverting 66a08027422e9cea9d6069bb6f2627b846813245 and  5bf2950a2b7f3b0510f189bf2d6c5b1ac924e3b1 brings back old behaviour (posted here as commit message makes finding this not so trivial).

**Comment by westnordost** [comment]
I only tested it with Chrome. Can you set a breakpoint? The logic that is executed when the user is led back to the app is line 144 (onNewIntent) in OsmOAuthDialogFragment. After that, line 159 is executed (continueAuthentication)

**Comment by krzyk** [comment]
For me in Firefox `continueAuthentication` is executed only once with state INITIAL (before the browser launch) and it never calls back that method with other states.

**Comment by westnordost** [comment]
and onNewIntent also does not trigger? Is perhaps onCancel triggered just when the browser is opened?

**Comment by Binnette** [comment]
I am affected too. Android browser is openned. I connect. But when it come back to SC, I still see "Not authorized yet".
Adnroid : 6.0
Browser : default android browser

**Comment by krzyk** [comment]
@westnordost onNewIntent nor onCancel is called.
When the browser closes (and redirects to streetcomplete:// url) it goes to `SettingsActivity.onCreate`, but `getIntent().getBooleanExtra(EXTRA_LAUNCH_AUTH, false))` returns false and doesn't start the `OsmOAuthDialogFragment` fragment.

I checked the intent data and it has the oauth url with oauth_verifier query param, but when I forced it to create the `OsmOAuthDialogFragment` it failed on `onCreate`, because the `Bundle` doesn't have the right data (e.g. inState is null).

I'll try to look at it further, but I'm not familiar with Android.

**Comment by westnordost** [comment]
What, onCreate?! I think that is the problem, Chrome doesn't do that.

I am not too proficient with Android tasks, but I think it has something to do with that Firefox is starting a new instance of SettingsActivity as a new task instead of reusing the SettingsActivity that was there before.

https://developer.android.com/guide/components/activities/tasks-and-back-stack.html

**Comment by krzyk** [comment]
OK, so is there a reason why you have `android:launchMode="singleTop"` for `SettingsActivity`?
Because based on the link above and the way I use SC it seams that SettingsActivity should have only a single instance always, not only when it is at the top.

Are there a cases when multiple SettingsActivities could be started?
If no, then a simple solution would be to change the android:launchMode="singleTop"` to `android:launchMode="singleTask"` or even `android:launchMode="singleInstance"`.

And with singleTask/singleInstance authenticating using Firefox (and Chrome) works fine.
Why it works with `singleTop` with Chrome? I have no idea :), probably they switch to other Activities in different way than Firefox/Andoird Browser.

I've checked Yandex browser, but it behaves same way as Chrome, no difference between `singleTop` and `singleTask`.

@Binnette is there a way to install the Android Browser from the Google Play? Or is it available only preinstalled?

**Comment by Binnette** [comment]
@krzyk: it seems to be possible. This default browser is called "Stock Browser" ou "AOSP Browser". Ask your favorite search engine about it.

On my device Homtom HT17 Pro Android 6.0. This is simply called "Browser" and version of the app is "6.0-1483499032".

If you can not install it. May be, you can try to install a virtual device on your computer. I hope it will be installed by default.

**Comment by krzyk** [comment]
@matkoniecz @Binnette Please check if this bug still exists in current master

**Comment by matkoniecz** [comment]
@krzyk Now it works for Firefox. 

It leaves open openstreetmap.org/oauth/authorize?oauth_token=XXXXX page that may be slighly confusing - but I am not sure is it something that may be changed.

**Comment by Binnette** [comment]
Bug no longer existe with la test master with thé "Android Stock Browser". Good job ! ☺

**Comment by westnordost** [comment]
Thank you @krzyk !
